### ISSUE TITLE
How to pad all the data into one single size instead of varying size?

### ISSUE THREAD

**Issue by ivicts** [atomistic-machine-learning/schnetpack#169]
Hi,

I want to pad all the data (not a single batch) into a single size - the size of the largest molecule in the data. So, for QM9, I want all the data's properties (atom_mask, atomic_numbers, positions, etc) referring to an atom size of 29 because it is the size of the largest molecule in the QM9. How do I do this?

Can I confirm that how you pad: In one batch size (100), you pick the largest molecule sizes and use that size for the whole batch?

**Comment by mgastegger** [comment]
Yes, we pick the dimensions of the largest molecule in each batch.
The maximum size is determined in lines 29-31 of the function `_collate_aseatoms` in  `src/schnetpack/data/loader.py`.

If you want to use it during training, you could try setting the `max_size` value to a static number, e.g. `max_size = 29`. If you just want to reformat the data, you can also use a `DataLoader` where you set the batch size equal to the length of the dataset.

**Comment by ivicts** [comment]
Hi,

max_size is actually a dictionary rather than a single value. This is one example of max_size:
`
{'energy_U0': array([1]), '_atomic_numbers': array([14]), '_positions': array([14,  3]), '_cell': array([3, 3]), '_neighbors': array([14, 13]), '_cell_offset': array([14, 13,  3]), '_idx': array([1])}
`
I have tried to changed the code for `_collate_aseatoms` in `src/schnetpack/data/loader.py` to this:

```
def _collate_aseatoms(examples):
    """
    Build batch from systems and properties & apply padding

    Args:
        examples (list):

    Returns:
        dict[str->torch.Tensor]: mini-batch of atomistic systems
    """
    properties = examples[0]

    # initialize maximum sizes
    max_size = {
        prop: np.array(val.size(), dtype=np.int) for prop, val in properties.items()
    }

    # # get maximum sizest i
    # for properties in examples[1:]:
    #     for prop, val in properties.items():
    #         max_size[prop] = np.maximum(
    #             max_size[prop], np.array(val.size(), dtype=np.int)
    #         )

    max_mol_size = 29
    max_size["_atomic_numbers"] = np.array([max_mol_size])
    max_size["_positions"] = np.array([max_mol_size, 3])
    max_size["_neighbors"] = np.array([max_mol_size, max_mol_size - 1])
    max_size["_cell_offset"] = np.array([max_mol_size, max_mol_size - 1, 3])

    # initialize batch
    # the rest of the codes is the same
```

It seems that this gives what I want.

1. Can you confirm that I only need to change "_atomic_numbers", "_positions", "_neighbors", "_cell_offset" key of the dictionary regardless of what dataset is used?
2. Can I confirm that this part of the codes is correct regardless of what dataset is used?
` 
    max_size["_atomic_numbers"] = np.array([max_mol_size])
    max_size["_positions"] = np.array([max_mol_size, 3])
    max_size["_neighbors"] = np.array([max_mol_size, max_mol_size - 1])
    max_size["_cell_offset"] = np.array([max_mol_size, max_mol_size - 1, 3])
`
3.  What do you mean by using it during training? Yes, I want the model to trained and tested on one size of matrix.
4. Can I confirm that this would not affect the other part of the codes?

**Comment by mgastegger** [comment]
1) Ah, yes I forgot. You basically need to change all entries containing the number of atoms as an dimension, which are exactly the ones you listed.
2) The sizes are correct, `_neighbors` has the dimension `n_atoms x n_neighbors`, which for the standard environment provider reduces to `n_atoms x n_atoms-1` (no double counting of the central atom). `_cell_offset` is similar.
3) This was just in case you only wanted to reformat the data for analysis.
4) It should not affect other parts of the code, as long as the `_atom_mask` and `_neighbor_mask` are set properly, which is done automatically in `_collate_aseatoms`. You could run into problems with this approach if you want to model properties which contain the number of atoms in their shape (e.g. forces with `n_atoms x 3`). However, everything in QM9 is fine.

**Comment by ivicts** [comment]
Hi, thanks for clarification for 1 to 3.

So, for (4), you meant that for properties which contain the number of atoms in their shape (e.g forces), I may need to pad the prediction label array as well is it?

Besides QM9, I may want to do MD17 and ISO17 dataset. As far as I know, they are having the same atom sizes for each task, so we will not need to pad this dataset right? So, this implementation should be also OK for MD17 and ISO17 dataset right?

**Comment by mgastegger** [comment]
Yes, exactly.

For the ISO17 dataset all molecules have the same number of atoms, so forces should be no problem. In the MD17 dataset it depends on whether you want to look on a single molecule (e.g. ethanol) or multiple molecules (e.g. ethanol and aspirin) at the same time. In the first case, there is no problem, in the second case you need to pad the forces.

**Comment by ivicts** [comment]
@mgastegger thanks for the kind help!
